lucene/core/src/java/org/apache/lucene/search/FieldValueChecker.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+/**
+ * Maintains bottom feature value across multiple collectors
+ */
+public class FieldValueChecker {
+  public volatile Object value;
+  public volatile int minimumDoc;
+  private FieldComparator[] fieldComparators;
+  private int[] reverseMul;
+  private boolean isEnabled = false;
+
+  public FieldValueChecker(final FieldComparator[] fieldComparators, final int[] reverseMul) {
+    assert fieldComparators.length == reverseMul.length;
+
+    this.fieldComparators = fieldComparators;
+    this.reverseMul = reverseMul;
+  }
+
+  void disableGlobalHitCheck() {
+    isEnabled = false;
+  }
+
+  boolean isBottomValuePresent() {
+    if (!isEnabled) {
+      return false;
+    }
+
+    return this.value != null;
+  }
+
+  void checkAndUpdateBottomValue(Object value, int doc) {
+    if (!isEnabled) {
+      return;
+    }
+
+    synchronized (this) {
+      if (this.value == null) {
+        this.value = value;
+        this.minimumDoc = doc;
+        return;
+      }
+    }
+
+    if (isValueCompetitive(value, doc)== true) {
```

Comment:
There is no synchronization here, so you can end up comparing a minimum doc that is not associated to the current value ?

Reply:
I am not too sure here -- the checks before this assignment are synchronized. However, I do agree that being safe is key here -- will update to include this in the synchronized block, thanks